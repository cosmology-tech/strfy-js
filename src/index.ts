export * from './js';
export * from './json';