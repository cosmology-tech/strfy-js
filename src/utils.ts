
export function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function escapeStringForSingleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only single quotes
  str = str.replace(/'/g, "\\'");
  return str;
}

export function escapeStringForDoubleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only double quotes
  str = str.replace(/"/g, '\\"');
  return str;
}


export function escapeStringForBacktickQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape backticks
  str = str.replace(/`/g, '\\`');
  return str;
}
