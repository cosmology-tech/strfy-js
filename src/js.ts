import { camelCaseTransform, escapeStringForBacktickQuotes, escapeStringForDoubleQuotes, escapeStringForSingleQuotes, isSimpleKey } from "./utils";

export interface JSStringifyOptions {
  space?: number; // defaults to 0
  replacer?: (this: any, key: string, value: any) => any | null; // defaults to null
  quotes?: 'single' | 'double'; // defaults to single
  inlineArrayLimit?: number; // default is undefined, allowing inline arrays below this length
  camelCase?: boolean; // defaults to false
  camelCaseFn?: (str: string) => string; // optional function to convert keys to camelCase
}

export function chooseQuotes(str: string, preferred: 'single' | 'double' | 'backtick'): string {
  switch (preferred) {
    case 'single':
      return `'${escapeStringForSingleQuotes(str)}'`;
    case 'double':
      return `"${escapeStringForDoubleQuotes(str)}"`;
    case 'backtick':
      return `\`${escapeStringForBacktickQuotes(str)}\``;
    default:
      throw new Error("Invalid quote type specified.");
  }
}


export function jsStringify(obj: any, options?: JSStringifyOptions): string {
  const { space = 0, replacer = null, quotes = 'single', inlineArrayLimit = undefined, camelCase = false, camelCaseFn = camelCaseTransform } = options || {};
  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);
  let indentLevel: number = 0;

  const serialize = (obj: any): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map(item => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];
        const transformedKey = camelCase ? camelCaseFn(key) : key;
        const keyPart = isSimpleKey(transformedKey) ? transformedKey : `"${transformedKey}"`;
        const valuePart = serialize(value);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return chooseQuotes(obj, quotes);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}