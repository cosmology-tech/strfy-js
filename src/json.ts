import { camelCaseTransform, isSimpleKey } from "./utils";

export interface JSONStringifyOptions {
  space?: number; // defaults to 0
  replacer?: (this: any, key: string, value: any) => any; // defaults to null
  quotes?: 'single' | 'double'; // defaults to double
  inlineArrayLimit?: number; // default is undefined, allowing inline arrays below this length
  camelCase?: boolean; // defaults to false
  camelCaseFn?: (str: string) => string; // optional function to convert keys to camelCase
}

function escapeString(str: string): string {
  str = str.replace(/\\/g, '\\\\');  // Escape backslashes first
  str = str.replace(/\n/g, '\\n');  // Escape newlines
  str = str.replace(/\r/g, '\\r');  // Escape carriage returns
  str = str.replace(/\t/g, '\\t');  // Escape tabs
  str = str.replace(/"/g, '\\"');  // Escape double quotes
  return str;
}

function chooseQuotes(str: string): string {
  return `"${escapeString(str)}"`;
}

export function jsonStringify(obj: any, options?: JSONStringifyOptions): string {
  const { space = 0, replacer = null, quotes = 'double', inlineArrayLimit = undefined, camelCase = false, camelCaseFn = camelCaseTransform } = options || {};
  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);
  let indentLevel: number = 0;

  const serialize = (obj: any): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map(item => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];
        const transformedKey = camelCase ? camelCaseFn(key) : key;
        const keyPart = isSimpleKey(transformedKey) ? transformedKey : chooseQuotes(transformedKey);
        const valuePart = serialize(value);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return chooseQuotes(obj);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}
