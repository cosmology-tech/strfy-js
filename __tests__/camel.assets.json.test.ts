import { jsonStringify } from "../src";

import assetList from '../__fixtures__/assets.json';

it('assetlist', () => {
  expect(jsonStringify(assetList, { camelCase: true, space: 2 })).toMatchSnapshot();
});