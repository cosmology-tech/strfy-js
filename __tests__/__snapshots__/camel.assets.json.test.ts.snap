// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assetlist 1`] = `
"{
  $schema: "../assetlist.schema.json",
  chainName: "comdex",
  assets: [
    {
      description: "Native Token of Comdex Protocol",
      denomUnits: [
        {
          denom: "ucmdx",
          exponent: 0
        },
        {
          denom: "cmdx",
          exponent: 6
        }
      ],
      base: "ucmdx",
      name: "Comdex",
      display: "cmdx",
      symbol: "CMDX",
      logoURIs: {
        png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
        svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
      },
      coingeckoId: "comdex",
      images: [
        {
          png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
          svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
        }
      ]
    },
    {
      description: "Governance Token of Harbor protocol on Comdex network",
      denomUnits: [
        {
          denom: "uharbor",
          exponent: 0
        },
        {
          denom: "harbor",
          exponent: 6
        }
      ],
      base: "uharbor",
      name: "Harbor",
      display: "harbor",
      symbol: "HARBOR",
      logoURIs: {
        png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png",
        svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg"
      },
      coingeckoId: "harbor-2",
      images: [
        {
          png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png",
          svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg"
        }
      ]
    },
    {
      description: "Stable Token of Harbor protocol on Comdex network",
      denomUnits: [
        {
          denom: "ucmst",
          exponent: 0
        },
        {
          denom: "cmst",
          exponent: 6
        }
      ],
      base: "ucmst",
      name: "CMST",
      display: "cmst",
      symbol: "CMST",
      logoURIs: {
        png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png",
        svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg"
      },
      coingeckoId: "composite",
      images: [
        {
          png: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png",
          svg: "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg"
        }
      ]
    }
  ]
}"
`;
